Make encode-marc21 validation optional or optionally less strict

Currently the marc encoder only serializes to marc21 if a lot of validations are true.

Hbz/Verbundgruppe has a use case where for the it is more important to change the form from marcxml to marc21 than having valid marc21.

[The transformation breaks if the data is not valid at the moment.](https://metafacture.org/playground/?flux=inputFile%0A%7C+open-file%0A%7C+decode-xml%0A%7C+handle-marcxml%28namespace%3D%22%22%29%0A%7C+encode-marc21%0A%7C+print%0A%3B&data=%3C%3Fxml+version%3D%221.0%22+encoding%3D%22UTF-8%22%3F%3E%0A%3Ccollection%3E%3Crecord%3E%3Cleader%3E00600ny++a22002053n+4500%3C/leader%3E%3Ccontrolfield+tag%3D%22001%22%3E22235669990006443%3C/controlfield%3E%3Ccontrolfield+tag%3D%22008%22%3E2011090u++++0+++0000uuger0000000%3C/controlfield%3E%3Ccontrolfield+tag%3D%22005%22%3E20210825133059.0%3C/controlfield%3E%3Ccontrolfield+tag%3D%22007%22%3Ecr+%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%3C/controlfield%3E%3Ccontrolfield+tag%3D%22003%22%3EDE-605%3C/controlfield%3E%3Cdatafield+tag%3D%22016%22+ind1%3D%227%22+ind2%3D%22+%22%3E%3Csubfield+code%3D%22a%22%3E2899355-X%3C/subfield%3E%3Csubfield+code%3D%222%22%3EDE-600%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22092%22+ind1%3D%22+%22+ind2%3D%22+%22%3E%3Csubfield+code%3D%22a%22%3E401000-0%3C/subfield%3E%3Csubfield+code%3D%22k%22%3ENRW%3C/subfield%3E%3Csubfield+code%3D%22l%22%3Ea%3C/subfield%3E%3Csubfield+code%3D%22d%22%3E61%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22852%22+ind1%3D%22+%22+ind2%3D%221%22%3E%3Csubfield+code%3D%229%22%3E00%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22856%22+ind1%3D%224%22+ind2%3D%22+%22%3E%3Csubfield+code%3D%22u%22%3Ehttps%3A//www.ulb.hhu.de/link/dbis.html%3Flett%3D%26amp%3Btitel_id%3D102421%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22866%22+ind1%3D%22+%22+ind2%3D%220%22%3E%3Csubfield+code%3D%22z%22%3EZugriff+nur+f%C3%BCr+autorisierte+Benutzer+nach+Registrierung+mit+universit%C3%A4rer+EMail-Adresse%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22852%22+ind1%3D%22+%22+ind2%3D%22+%22%3E%3Csubfield+code%3D%22a%22%3EDE-61%3C/subfield%3E%3C/datafield%3E%3C/record%3E%3Crecord%3E%3Cleader%3E00334ny++a22001213n+4500%3C/leader%3E%3Ccontrolfield+tag%3D%22001%22%3E22350703640006443%3C/controlfield%3E%3Ccontrolfield+tag%3D%22008%22%3E2203240u++++8+++4001uuger0000000%3C/controlfield%3E%3Ccontrolfield+tag%3D%22005%22%3E20240103130309.0%3C/controlfield%3E%3Ccontrolfield+tag%3D%22003%22%3EDE-605%3C/controlfield%3E%3Cdatafield+tag%3D%22016%22+ind1%3D%227%22+ind2%3D%22+%22%3E%3Csubfield+code%3D%22a%22%3E204679-9%3C/subfield%3E%3Csubfield+code%3D%222%22%3EDE-600%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22092%22+ind1%3D%22+%22+ind2%3D%22+%22%3E%3Csubfield+code%3D%22a%22%3E401000-0%3C/subfield%3E%3Csubfield+code%3D%22d%22%3E61%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22852%22+ind1%3D%22+%22+ind2%3D%221%22%3E%3Csubfield+code%3D%22c%22%3ELS+01%3A+Auslage+alg%3C/subfield%3E%3Csubfield+code%3D%229%22%3E00%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22866%22+ind1%3D%223%22+ind2%3D%220%22%3E%3Csubfield+code%3D%22a%22%3ELaufender+Jahrgang%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22852%22+ind1%3D%22+%22+ind2%3D%22+%22%3E%3Csubfield+code%3D%22a%22%3EDE-61%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22852%22+ind1%3D%22+%22+ind2%3D%222%22%3E%3Csubfield+code%3D%22b%22%3ELS%3C/subfield%3E%3Csubfield+code%3D%229%22%3E09%3C/subfield%3E%3C/datafield%3E%3Cdatafield+tag%3D%22852%22+ind1%3D%22+%22+ind2%3D%22+%22%3E%3Csubfield+code%3D%228%22%3E2%3C/subfield%3E%3C/datafield%3E%3C/record%3E%3C/collection%3E)

I therefore suggest to make the validation of the leader and controllfields optional.

> Note that without setting the `namespace` option like in `| handle-marcxml(namespace="")` nothing will happen (as in your example).

This is only partly true:

```
$ '/home/tobias/Downloads/metafix-runner-1.1.2/bin/metafix-runner' '/home/tobias/Downloads/Unbenannter Ordner/playground.flux' 
Exception in thread "main" java.lang.AssertionError: unknown or unexpected state: IN_STREAM
	at org.metafacture.biblio.marc21.Marc21Encoder.literal(Marc21Encoder.java:195)
	at org.metafacture.biblio.marc21.MarcXmlHandler.endElement(MarcXmlHandler.java:135)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.endElement(AbstractSAXParser.java:618)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEndElement(XMLDocumentFragmentScannerImpl.java:1728)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2899)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
	at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:73)
	at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:44)
	at org.metafacture.io.FileOpener.process(FileOpener.java:158)
	at org.metafacture.io.FileOpener.process(FileOpener.java:41)
	at org.metafacture.flux.parser.StringSender.process(StringSender.java:43)
	at org.metafacture.flux.parser.Flow.start(Flow.java:118)
	at org.metafacture.flux.parser.FluxProgramm.start(FluxProgramm.java:168)
	at org.metafacture.runner.Flux.main(Flux.java:87)
```

--- a/metafacture-biblio/src/main/java/org/metafacture/biblio/marc21/Marc21Encoder.java
+++ b/metafacture-biblio/src/main/java/org/metafacture/biblio/marc21/Marc21Encoder.java
@@ -81,6 +81,7 @@ public final class Marc21Encoder extends
     private State state = State.IN_STREAM;
 
     private boolean generateIdField;
+    private boolean validateLeader = true;
 
     /**
      * Initializes the encoder with MARC 21 constants and charset.
@@ -106,6 +107,18 @@ public final class Marc21Encoder extends
      */
     public void setGenerateIdField(final boolean generateIdField) {
         this.generateIdField = generateIdField;
+    }
+
+    /**
+     * Controls whether the leader should be validated.
+     * <p>
+     * The default value of {@code validateLeader} is true.
+     * <p>
+     *
+     * @param validateLeader if false the leader is not validated
+     */
+    public void setValidateLeader(final boolean validateLeader) {
+        this.validateLeader = validateLeader;
     }
 
     /**
@@ -259,12 +272,14 @@ public final class Marc21Encoder extends
     }
 
     private void requireValidCode(final char code, final char[] validCodes) {
-        for (final char validCode: validCodes) {
-            if (validCode == code) {
-                return;
+        if (validateLeader) {
+            for (final char validCode : validCodes) {
+                if (validCode == code) {
+                    return;
+                }
             }
+            throw new FormatException("invalid code '" + code + "'; allowed codes are: " + Arrays.toString(validCodes));
         }
-        throw new FormatException("invalid code '" + code + "'; allowed codes are: " + Arrays.toString(validCodes));
     }
 
     private void processTopLevelLiteral(final String name, final String value) {

--- a/metafacture-biblio/src/test/java/org/metafacture/biblio/marc21/Marc21EncoderTest.java
+++ b/metafacture-biblio/src/test/java/org/metafacture/biblio/marc21/Marc21EncoderTest.java
@@ -38,6 +38,8 @@ import org.mockito.MockitoAnnotations;
  */
 public final class Marc21EncoderTest {
 
+    private static final String BAD_LEADER = "00600ny  a22002053n 4500";
+
     private Marc21Encoder marc21Encoder;
 
     @Mock
@@ -145,6 +147,23 @@ public final class Marc21EncoderTest {
         marc21Encoder.endRecord();
 
         verify(receiver).process(matches("00055pam a2200037 c 4500021001700000\u001e.*\u001d"));
+    }
+
+    @Test(expected = FormatException.class)
+    public void issue567ShouldFailValidateLeaderAsDefault() {
+        marc21Encoder.startRecord("");
+        marc21Encoder.literal(LEADER_ENTITY, BAD_LEADER);
+        marc21Encoder.endRecord();
+    }
+
+    @Test
+    public void issue567ShouldNotValidateLeader() {
+        marc21Encoder.setValidateLeader(false);
+        marc21Encoder.startRecord("");
+        marc21Encoder.literal(LEADER_ENTITY,   BAD_LEADER );
+        marc21Encoder.endRecord();
+
+        verify(receiver).process(matches("00026ny  a22000253n 4500\u001e\u001d"));
     }
 
 }

